### DISCUSSION TITLE
How to get history of relationship when associated id column get updated ?

### DISCUSSION THREAD

**Discussion by ikus060** [sqlalchemy/sqlalchemy#8733] [Usage Questions] [answered]
Hello,

I have a scenario where I want to track the history of a relationship. For that, I want to use the sqlalchemy history. But it seams only the id column get track and relationship is not track...

e.g.:
```py
from sqlalchemy import *
from sqlalchemy.orm import *
from sqlalchemy.ext.declarative import declarative_base

Base = declarative_base()

class Person(Base):
    __tablename__ = 'person'
    id = Column(Integer(), primary_key=True)
    name = Column(String(80))
    phone_id = Column(Integer(), ForeignKey("phone_number.id"))
    phone = relationship("PhoneNumber", active_history=True)

class PhoneNumber(Base):
    __tablename__ = 'phone_number'
    id = Column(Integer(), primary_key=True)
    type = Column(String(20))
    number = Column(String(11))

e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

session = Session(e)

p = Person()
p.name = "Frank"
phone1 = PhoneNumber(type="Mobile", number="1234567890")
phone2 = PhoneNumber(type="Home", number="9876543210")
session.add_all([phone1, phone2])
session.commit()
p.phone_id = phone1.id
session.add(p)
session.commit()
p = session.query(Person).first()
assert p.phone_id == phone1.id
p.phone_id = phone2.id

history = inspect(p).attrs.get("phone_id").history
assert history == ([phone2.id], (), [phone1.id])

history = inspect(p).attrs.get("phone").history
assert history == ([phone2], (), [phone1])
```

I would expect "phone" attribute to also get a history but it doesn't.

**Comment by zzzeek** [comment, marked as answer]
the attrs.history interface only tracks explicit user-initiated attribute set operations, and is additionally only present up until the next flush.  In your example above, you've assigned a value to `p.phone_id`, so that's an attribute set which you can see in .history, however you didn't do anything to `p.phone`, which hasn't been changed at all.   The way that you'd see `p.phone` returning a new object would be after the session is flushed, and then after the `p.phone` attribute were either expired, or if it were never loaded in the first place, it would lazy load the new PhoneNumber object based on the `p.phone_id` that was flushed to the database.

There's a little bit of background on this general idea at https://docs.sqlalchemy.org/en/14/faq/sessions.html#i-set-the-foo-id-attribute-on-my-instance-to-7-but-the-foo-attribute-is-still-none-shouldn-t-it-have-loaded-foo-with-id-7 , but beyond that "history" is not tracking  a net change in the effective value of attributes, only events in which the user modified the value directly.

the interesting thing here is that since in your example "phone" is unloaded (was expired on previous commit()), if you use [load_history()](https://docs.sqlalchemy.org/en/14/orm/internals.html#sqlalchemy.orm.AttributeState.load_history) , that actually fires off loader callables, which would fire off autoflush and you'd see the new "phone" object, but it wouldnt be stored as a change, it would be the current value:

```py
history = inspect(p).attrs.get("phone").load_history()

assert history == ((), [phone2], ())
```

**Comment by ikus060** [comment]
> the attrs.history interface only tracks explicit user-initiated attribute set operations,

ok, I will then need to remedy to this with custom code I guess.

Thanks for the clarification.

**Reply by zzzeek** [reply]
there's a little bit of unclear / not exactly accurate wording in that FAQ entry about "initialized", will update that now
